benches/nexmark/main.rs

```diff
@@ -2,64 +2,122 @@
 #![feature(is_some_with)]
+use std::{cell::Cell, rc::Rc, time::Instant};
+
 use anyhow::Result;
 use clap::Parser;
 use dbsp::{
-    circuit::Root,
+    circuit::{Circuit, Root},
     nexmark::{
-        config::Config as NexmarkConfig, generator::config::Config as GeneratorConfig,
-        model::Event, queries::q1, NexmarkSource,
+        config::Config as NexmarkConfig,
+        generator::config::Config as GeneratorConfig,
+        model::Event,
+        queries::{q0, q1, q2, q3, q4},
+        NexmarkSource,
     },
-    profile::CPUProfiler,
     trace::{ord::OrdZSet, BatchReader},
 };
 use rand::prelude::ThreadRng;
 
-// TODO: Enable running specific tests
+// TODO: Ideally these macros would be in a separate `lib.rs` in this benchmark
+// crate, but benchmark binaries don't appear to work like that (in that, I
+// haven't yet found a way to import from a `lib.rs` in the same directory as
+// the benchmark's `main.rs`)
+
+/// Returns a closure for a circuit with the nexmark source that sets
+/// `max_events_reached` once no more data is available.
+macro_rules! nexmark_circuit {
+    ( $q:expr, $generator_config:expr, $max_events_reached:expr ) => {
+        |circuit: &mut Circuit<()>| {
+            let source =
+                NexmarkSource::<ThreadRng, isize, OrdZSet<Event, isize>>::new($generator_config);
+            let input = circuit.add_source(source);
+
+            let output = $q(input);
+
+            output.inspect(move |zs: &OrdZSet<_, _>| {
+                // Turns out we can't count events by accumulating the lengths of
+                // the `OrdZSet` since duplicate bids are not that difficult to produce in the
+                // generator (0-3 per 1000), which get merged to a single Item in the `OrdZSet`
+                // with an adjusted weight. Instead, stop when we get empty sets.
+                // TODO(absoludity): Nope, can't do that either as some streams
+                // return empty sets for initial data (eg, q2 doesn't emit data until the 123rd
+                // auction, which is well beyond 1000 generated events).
+                println!("zset: {:?}", zs);
+                if zs.len() == 0 {
+                    $max_events_reached.set(true);
+                }
+            });
+        }
+    };
+}
+
+macro_rules! run_query {
+    ( $q:expr, $generator_config:expr ) => {{
+        // Until we have the I/O API to control the running of circuits,
+        // use an Rc<Cell<bool>> to communicate when the test is finished (ie. all
+        // events processed).
+        let max_events_reached = Rc::new(Cell::new(false));
+        let max_events_reached_cloned = max_events_reached.clone();
+
+        let circuit = nexmark_circuit!($q, $generator_config, max_events_reached);
+
+        let root = Root::build(circuit).unwrap();
+
+        let start = Instant::now();
+        loop {
+            if max_events_reached_cloned.get() {
+                break;
+            }
```

Comment:
Can't this be a while loop?

Reply:
Yep. I'd initially thought that putting the condition outside the loop (ie. the while condition) would mean that it wouldn't go out of scope between iterations, but seems fine, thanks.